x/gamm/keeper/migrate.go

```diff
@@ -13,56 +13,44 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-func (k Keeper) Migrate(ctx sdk.Context, sender sdk.AccAddress, sharesToMigrate sdk.Coin, poolIdEntering uint64) (amount0, amount1 sdk.Int, liquidity sdk.Dec, poolIdLeaving uint64, err error) {
+// MigrateFromBalancerToConcentrated migrates unlocked lp tokens from a balancer pool to a concentrated liquidity pool.
+// Fails if the lp tokens are locked (must utilize UnlockAndMigrate function in the superfluid module)
+func (k Keeper) MigrateFromBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, sharesToMigrate sdk.Coin) (amount0, amount1 sdk.Int, liquidity sdk.Dec, poolIdLeaving, poolIdEntering uint64, err error) {
 	// Get the balancer poolId by parsing the gamm share denom.
-	poolIdLeaving, err = getPoolIdFromSharesDenom(sharesToMigrate.Denom)
-	if err != nil {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
-	}
+	poolIdLeaving = types.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
 
-	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
-	migrationInfo := k.GetMigrationInfo(ctx)
-	matchFound := false
-	for _, info := range migrationInfo.BalancerToConcentratedPoolLinks {
-		if info.BalancerPoolId == poolIdLeaving {
-			if info.ClPoolId != poolIdEntering {
-				return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, types.InvalidPoolMigrationLinkError{PoolIdEntering: poolIdEntering, CanonicalId: info.ClPoolId}
-			}
-			matchFound = true
-			break
-		}
-	}
-	if !matchFound {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, types.PoolMigrationLinkNotFoundError{PoolIdLeaving: poolIdLeaving}
+	// Find the governance sanctioned link between the balancer pool and a concentrated pool.
+	poolIdEntering, err = k.GetLinkedConcentratedPool(ctx, poolIdLeaving)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, err
 	}
 
 	// Get the concentrated pool from the message and type cast it to ConcentratedPoolExtension.
 	poolI, err := k.clKeeper.GetPool(ctx, poolIdEntering)
 	if err != nil {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, err
 	}
 	concentratedPool, ok := poolI.(cltypes.ConcentratedPoolExtension)
 	if !ok {
 		// If the conversion fails, return an error.
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
```

Comment:
Agreed, did this with `CreateFullRangePosition` and `GetPoolFromPoolIdAndConvertToConcentrated`. While I could have actually put the `GetPoolFromPoolIdAndConvertToConcentrated` logic into `CreateFullRangePosition`, I want to have the poolId check in the very beginning to prevent overhead when running in invalid tx.

https://github.com/osmosis-labs/osmosis/pull/4129/commits/788065d31ce523e812cc679013c9dbde29ea6458